### ISSUE TITLE
Dockerfile: Shrinking the image from 1.6 GB to 600 MB with Debian Slim and multi-stage builds

### ISSUE THREAD

**Issue by nickjj** [rails/rails#46855]
Hi,

Currently the `Dockerfile` is referencing `FROM ruby:$RUBY_VERSION`. By default
it will use the full version of Debian.

- `rails new myapp` out of the box will produce a Docker image that is ~1.6 GB
- If you add flags that result in a Node environment being included the image is ~1.8 GB

You can verify this by running `docker image ls` after building the image(s).

I think we can substantially improve this. I've worked with a few Rails
projects in Docker where the baseline image size (including Node) is ~475 MB
while still using Debian.

#### There's at least 2 things we can do to reduce the size of the image:

- Switch to using Debian slim which is a variant of Debian that has hundreds of less pre-installed packages, no man pages and other reductions
- Start using multi-stage builds
  - Phase 1: Avoid including the entire Node environment into the main Rails app's stage
  - Phase 2: Potentially differentiate build time and run-time dependencies to further reduce image sizes

# Debian slim

Debian installs 528 `apt` packages where as Debian slim installs 211 packages.
I verified that by running `apt list --installed` inside of each container
after building them.

The main trade off by going with Debian slim is you may need to track down and
install certain `apt` packages to get certain gems to compile their native
dependencies. You win by having a smaller image size.

## A couple of packages you may need to install

The good news is gems like `bcrypt`, `openssl`, `iconv` and `sqlite3` work out of
the box with slim as long as you include `build-essential` which includes a
number of popular build tools and libraries.

For database access you'll get compile errors without these `apt` packages installed:

- `pg` requires `libpq-dev`
- `mysql2` requires `libmariadb-dev-compat`

## Image sizes with various packages

The first 4 image sizes below do not include Node because this has too much
variance based on what you plan to Yarn install. The Node run-time in itself is a
little over +200 MB but that's without Yarn installing any packages, this
number will grow substantially as you add JS packages.

| Description | Size        |
| ----------- | ----------- |
| Baseline `rails new myapp` | 1.6 GB |
| Baseline `rails new myapp` (removed `libvips`) | 1.3 GB |
| Slim with `build-essential curl git libvips` | 1.3 GB |
| Slim with `build-essential curl git` | 862 MB |
| [An example using slim + multi-stage builds](https://github.com/nickjj/docker-rails-example) | 475 MB |
| [An example using slim + multi-stage builds](https://github.com/nickjj/docker-rails-example) (with `libvips`) | 600 MB |

I know that last example app isn't the `rails new` generated `Dockerfile`. I'm
not trying to promote that project. I just wanted to include a reproducible
example that's open source and works.

That example app includes all of the Hotwire dependencies, `pg`, `sidekiq`,
`rack-mini-profiler` and a number of Node packages (`esbuild`, `tailwindcss`,
etc.) which means the base size is quite a bit larger due to there being extra
packages installed vs `rails new`.

*If I adjust my example app to not use Slim it goes from 475 MB to 960 MB.
That's a big difference!*

### Image processing with libvips (Active Storage dependency)

This is a system level Active Storage dependency.

It's interesting that it adds +300 MB to the non-slim image and +450 MB to the
slim version. When I add `libvips` to my project's example, oddly enough it
only adds +130 MB (it's ~600 MB grand total).

It's a beast of a package that installs a bunch of other packages.

### What about Postgres and MySQL?

Thankfully `libpq-dev` and `libmariadb-dev-compat` only add +1 MB each!

## Takeaway

With the current `Dockerfile` as is, switching to slim will drop things from
1.6 GB to about 1.3 GB which is about a 20% savings. That's not too bad.

I think this has potential to become a much bigger factor when combined with
multi-stage builds.

# Multi-stage builds

The basic idea is that you can perform certain actions in 1 stage and copy the
files it produces into another stage. All of this occurs in a single
`Dockerfile`.

### Phase 1: Splitting out the Node environment

This is great for let's say setting up a full Node environment so you can
pre-compile all of your assets and then only copy the final bundled assets over
to your main app's stage. At the end of the day Rails only cares that you have
assets supplied in the `public/` directory, it doesn't care how they got there.

It means you don't need to lug around a 200 MB Node run-time and then another
few hundred megs of `node_modules/`. Instead you can copy over a few CSS, JS,
images, fonts and other static files over.

### Phase 2: Separating build time and run-time system dependencies

The `build-essential` package includes many tools and libraries. There's a
number of gems that require these files to compile native dependencies but once
the compile phase is done, you may only need a statically linked binary to
exist so the gem can use it at run-time.

All in all you could have a build stage to handle installing your gems and then
another stage would copy over the bundled directory and required binaries into
another stage where your app runs.

I haven't delved down this rabbit hole yet, but I think it's worth exploring.

In my example project I do copy over the `/usr/local/bundle/` directory between
the Node stage into the main app stage -- this method does work.

In my case I did that because the Node stage needs to bundle install everything
to pre-compile the assets. I found in development copying over the bundle
directory was faster, especially since I spawn a `web`, `worker` and `cable`
container all from the same Docker image with different commands.

## Takeaway

I'd suggest rolling with stage 1 to begin with, get something working and then
consider stage 2.

This will be an interesting problem to solve at the generator (Thor template)
level because we may need quite a few `using_node?` checks to produce a clean
looking Dockerfile.

# Action items and open questions

Before opening any PRs to tackle some of the above I wanted to drop all of this
here to gauge interest and opinions from the Rails core team and others.

If we're striving for a production ready Dockerfile in my opinion file size is
decently important. It's much nicer shipping around a 400 MB image vs 2 GB
image.

I say "decently" because in a number of cases such as single server deploys or
self managed Kubernetes nodes your images will remain on the server and you
only need to pull diffs for each change but even so, having a smaller image is
preferable for initial pulls on new nodes or any Fargate-like environment where
images aren't cached.

The cost here is user complexity, end users will need to know about concepts
like multi-stage builds and maybe chase down system packages if one of their
gems fails to compile.

In my opinion chasing down packages is no problem. It's 1 Google search away
from finding an answer but to be fair that's after you know what the root cause
is (ie. you have experience). Then again, we are targeting production systems
right? I think we can't avoid all complexity.

As a separate matter to this issue, I'd also like to see running all processes
as a non-root user. My example app linked earlier supports that but it
certainly adds complexity to the Dockerfile since now we're dealing with
creating a user and setting a UID / GID.

What do you think?

**Comment by simi** [comment]
I think it would be super nice to have reference Rails production graded Rails Dockerfile image you can extend (following provided instructions to keep it slim as possible).

**Comment by jankeesvw** [comment]
> I think it would be super nice to have reference Rails production graded Rails Dockerfile image you can extend (following provided instructions to keep it slim as possible).

I would love to read a Rails guide about Docker.

**Comment by lazaronixon** [comment]
@nickjj https://community.fly.io/t/preparations-for-rails-7-1/9512/2

**Comment by nickjj** [comment]
@lazaronixon Thanks. DHH and I did chat on Discord around the time that thread was created but that's the first time I'm seeing the thread. I think we can address a number of his concerns, especially around using Slim and making sure common gems along with SQLite, Postgres and MySQL work without users needing to modify the Dockerfile. This is a pretty minor change IMO.

SQLite works out of the box with Slim + `build-essential` and both of other packages for `pg` and `mysql2` are only 1MB each so including them by default feels ok? A more experienced user can remove the one they're not using if they want to optimize their Dockerfile.

The multi-stage build aspect feels a bit more up to the air based on the replies in that thread. I've been using multi-stage builds for Node for years (in dev and prod) but that thread is talking about a lot of potential changes. Kind of a combination between phase 1 and phase 2 described in this issue. A decision should probably be made before any work on that moves forward?

**Comment by eileencodes** [comment]
The dockerfile in this directory is for running our CI and wasn't intended for production usage. The dockerfile for making a Rails app is over in https://github.com/rails/docked so I think this conversation should be moved there.

**Comment by louim** [comment]
@eileencodes I think this issue is referencing the newly added [Dockerfile generator](https://github.com/rails/rails/pull/46762) that will come with rails 7.1, not the static Dockerfile in the repo. From my understanding, Docked is intended be used for development.

**Comment by eileencodes** [comment]
Ah ok, well I've been out for two weeks so I missed that. Thanks for the link.

More generally I think if there are improvements to be made then PRs should be opened and we can discuss each change individually. Long running issues have a tendency to get ignored and eventually closed. I don't think most of us on Core are docker experts so getting an opinion without seeing code might be hard. 😄

**Comment by byroot** [comment]
> I don't think most of us on Core are docker experts so getting an opinion without seeing code might be hard.

I dealt with enough docker image builds at Shopify that I think I can take this one :)

So @nickjj 

> rails new myapp out of the box will produce a Docker image that is ~1.6 GB

Indeed, and it's true that all things equal, a smaller image will lead to better performance. That being said, I'd like to point a few things.

The size of the base layers don't matter nearly as much as the size of the later layers assuming they don't change too often as you pointed. Trading 600MB in a base layer that only change monthly or so, for an extra 100MB into a later layer that is built much more often can lead to worse performance.

Also I think it's important to point what this Dockerfile is. It's a template, it's copied into your app when you initially create it, after that the developer owns it. They can make all the changes and optimizations they feel confident doing, but the goal from the core team point of view is maximum compatibility as to give an easy, problem-free, quick-start.

That is why `alpine` images weren't considered. Now `ubuntu-slim` isn't nearly as bare, so it might be a good fit. Would be interesting to see the list of the packages that have been stripped to try to evaluate how likely it is to cause people to run into various build issues.

This also mean that, as you point out, we'd rather not add to much complexity for the users, so that people with limited knowledge run into less problems and can debug them easier.

And overall I think people who feel like optimizing further can easily use Dockerfile they find in blog posts and such. So my personal opinion is that I wouldn't change the Dockerfile much at this point.

But if you feel strongly about this, I'd suggest to open some small, focused PRs and we can more deeply discuss the merits of specific changes there.

On this note I'll close this issue as I don't think there is an open problem to address, and further discussion can happen either on PRs or on this closed issue.

**Comment by nickjj** [comment]
> Would be interesting to see the list of the packages that have been stripped to try to evaluate how likely it is to cause people to run into various build issues.

This issue does list how to do that and includes a breakdown of common gems that require compiling native dependencies that do work out of the box with `-slim`.

> But if you feel strongly about this, I'd suggest to open some small, focused PRs and we can more deeply discuss the merits of specific changes there.

Understood. Beyond what's listed in this issue, what merits are you looking to see covered? I think this issue breaks down the pros and cons and also demonstrates it's possible to reduce the image size by 1 full GB.

In my opinion the PRs weren't ready for implementation yet because there's a number of unanswered questions, especially around how to handle using Node vs not using Node. This could drastically change how it's implemented and save a lot of time if these things were signed off on before the work started.

> And overall I think people who feel like optimizing further can easily use Dockerfile they find in blog posts and such. 

Personally I don't think this will be the case. For someone to dive down this rabbit hole will require having a lot of Docker experience to put together a fully working optimized solution. There may also be a lot of unknown unknowns for them to search for.

As someone who has been using Docker since ~2014 and created a few Docker courses the conclusion I've came to is that most folks don't deeply care about Docker in isolation. They care about the end result of what Docker provides them.

Most want to teeter on the line of knowing a little bit more than just enough to be dangerous. They mainly want to either run or deploy their apps, specifically their Rails apps. Docker can help a lot here but for most, the enjoyment doesn't come from spending 100 hours combing through a dozen blog posts and trying to make sense of a bunch of different Docker concepts. The enjoyment comes from deploying their app and letting the world test their ideas.

Similarly the same can be said for front-end build tools right? I think the transition to using import maps and js-bundling/css-bundling are a huge step forward because not needing a PhD in Webpacker to make your apps look and feel good is great. Docker feels sort of related to this. It's an ancillary and important tool, but the main event is your Rails app.

> So my personal opinion is that I wouldn't change the Dockerfile much at this point.

Is this potentially a marketing change instead of a code change scenario?

At the moment the marketing for this Dockerfile is that it's production ready. Production ready would be optimized and handle a bunch of edge cases based on lots of experience. It would be a near drop-in solution that you can trust is correct because "it's Rails".

Also, having a 1.6 GB official Rails image (or ~2GB+ if you add Node) is adding more fuel to the fire for the naysayers who think Rails is too slow and bloated. _We_ know that's not the case but it's 1 more data point to say "XYZ framework's Docker image is 300MB, why is the official production ready Rails solution almost 10x that?".

With that said, if the marketing were changed to not be a production Dockerfile then if it stayed as is with no further changes folks would be more forgiving because the expectation wouldn't be a "Rails blessed" production solution.

Think about this from another angle. Imagine generating a Rails app and using ActiveRecord but the expectation would be that to really use it in production users would need to research how to make it production ready on their own with no other guidance. I think there's an expectation that what Rails provides is a really good experience. If Rails says it's production ready, it should live up to that. That's the Rails standard. If that's not the case here, then it tarnishes the brand as a whole.

Here's just a couple of issues folks will potentially run into using this Dockerfile with no further changes:

- The container runs all processes as the root user
  - Even with it running in a container this poses security risks
- The image is +1 GB larger than it could be
  - A number of container platforms doesn't cache images so you'll be pulling this on every change
  - It drastically increases surface area of what can be taken advantage of through exploits and vulnerabilities
- Docker Compose isn't included which means while their app is running in Docker, users need to figure out how to run Postgres or MySQL and Redis on their own
- The `database.yml` that's generated isn't compatible with how the official Docker Postgres and MySQL images work
  - They might use a hosted database in production but for smaller scale 1 server deploys it's not unreasonable to self host your DB
- They'll be left on their own trying to serve assets in production and may run into an assortment of issues
  - One example will be what happens when you have multiple sprocket manifest files when you use volumes

The above focuses specifically on production. There's a number of other issues and pitfalls in development and the other `docked` project doesn't solve all of them.

Understandably development is a whole different story. The Docker experience is great with native Linux and WSL 2 as long as your code is in WSL 2 file's system but in large Rails apps the file system performance on macOS is pretty bad. I understand that might be why Rails isn't targeting Docker for development by default.

Again, I'm not trying to self promote anything but https://github.com/nickjj/docker-rails-example/ is a working example that addresses all of the production issues listed above and is also equipped to run in development if you're ok with Docker's dev performance. There's full dev / ci / prod parity with a single `docker compose up --build` command. It's there to demonstrate you can produce an optimized solution that "just works".

I'm happy to contribute more PRs. For example switching to slim is pretty easy and mostly non-invasive, but multi-stage builds when accounting for generator options to have Node or not is pretty big. If members want to leave it as is, that's cool too. The messaging isn't clear as things currently stand.

**Comment by byroot** [comment]
> > And overall I think people who feel like optimizing further can easily use Dockerfile they find in blog posts and such.
> 
> Personally I don't think this will be the case. For someone to dive down this rabbit hole will require having a lot of Docker experience to put together a fully working optimized solution.

I wasn't suggesting that users will read posts to become Docker experts, but that they'll find avrious advices such as switching to `-slim` images to slightly optimize their Dockerfile.

> At the moment the marketing for this Dockerfile is that it's production ready. Production ready would be optimized and handle a bunch of edge cases based on lots of experience. It would be a near drop-in solution that you can trust is correct because "it's Rails".

Not sure what you are referring to. The intent here is to have something that will work out of the box for the vast majority of users until they get to a point where they need something more tailored to their need. It's the usual rule of the framework: make it work by default without any config for 80%+ of the users. People who need something more specific will have to configure or customize.

The goal is to allow to deploy very quickly to a bunch of platforms.

> the enjoyment doesn't come from spending 100 hours combing through a dozen blog posts and trying to make sense of a bunch of different Docker concepts.

Yes. Exactly why I'd rather keep the Dockerfile as simple as possible so that they're less likely to run into build or runtime issues. 

multi-stage is typically the advanced used that can trip users. e.g. a gem that compiles because the dependency is there is the second stage, but fail to load because it's missing in the final image.

Same with node. Ideally you shouldn't need it in the final image, so you could save a lot by not including it, but then many users don't understand why they can't invoke it.

> Also, having a 1.6 GB official Rails image (or ~2GB+ if you add Node) is adding more fuel to the fire for the naysayers

Again, the overall size of an image isn't a good indication of performance. I can't help that there are morons on the internet...

> The container runs all processes as the root user 

I'd be open to a PR creating a user. The problem with this is permissions across mounted volumes, but since this Dockerfile is intended for production, it shouldn't be too much of an issue. That's generally more of an issue for development / CI.

> The image is +1 GB larger than it could be 

Again. I'm open to `-slim` images, but it requires a bit of research first to see what common packages may be missing. e.g. trying to install some common open source apps to see what they depend on. Just `pg` and `mysql` likely isn't enough.

> * Docker Compose isn't included

This Dockerfile is meant for deployment, not for development. So docker-compose doesn't make sense. The various datastores are expected to be provided by the hosting platform nad exposed via environment variables.

**Comment by nickjj** [comment]
> Not sure what you are referring to. 

The original PR from DHH at https://github.com/rails/rails/pull/46762 mentioned it being focused on a production deploy for the application. To be fair he did say "starting point" but I think using terms like production has a lot of weight to it. If you're expected to change half a dozen things, it's still technically a starting point but it leaves a lot on the user to figure out.

> multi-stage is typically the advanced used that can trip users. e.g. a gem that compiles because the dependency is there is the second stage, but fail to load because it's missing in the final image.

This issue broke up multi-stage builds into 2 phases. One for splitting out node and one for splitting out build vs runtime dependencies. I do agree it's trickier to test and debug splitting out build time vs run-time dependencies. This issue has a whole section dedicated to that being probably not worth it in the end -- at least to begin with. That's under the "Phase 2: Separating build time and run-time system dependencies" section.

> Again, the overall size of an image isn't a good indication of performance

It won't affect application run-time performance but it does affect rolling out new versions if you're using a container service that doesn't cache images to disk. Pulling an extra 1+ GB on every deploy x N replicas over the network takes time.

I've used EKS Fargate in the past and it's notorious for this, as is ECS Fargate. A number of cloud based services that are creating abstractions on top of Kubernetes also don't persist images. You get a fresh "machine" on every deploy. This may improve over time as those systems mature but current day is pretty oriented towards a world where having smaller images is a benefit within reason of course. For example, in most cases it's not worth making a crazy complicated solution to shave 50 MB off a 500 MB image, but shaving off 1 GB is a lot.

> I'd be open to a PR creating a user. The problem with this is permissions across mounted volumes, but since this Dockerfile is intended for production, it shouldn't be too much of an issue. That's generally more of an issue for development / CI.

This has implications in all environments.

**In development**, Docker Desktop handles mapping permissions so even if your container runs as `root:root`, volume mounted files on disk will be owned by your user/group. But if you're running native Linux or WSL 2 without Docker Desktop then files will be owned by `root:root` which is a very big inconvenience, especially when using any Rails generator or running a `db:migrate` since these commands produce files as the root user through a volume. You can't even edit those files with your editor without escalating privileges!

**In production**, it's the same story as native Linux. You'll end up with files owned by `root:root`.

A PR to add this feature will introduce a fair amount of complexity because by default Docker will map permissions to the first user in the system based on their UID / GID which will be 1000:1000. For bind mounts, the UID / GID will be used for mapping permissions back to the Docker host.

Using volumes in production is very common for single server deploys, specifically for bind mounting your `./public` directory so a web server like nginx can serve it as its root location. I can't count the number of set ups I've seen over the years that do this as a contract worker, this is also the approach I take as well. nginx runs directly on your Docker host, there's a lot of benefits for this that go beyond the scope of this issue but I'm happy to elaborate on it if you want.

In production, if your user'd UID / GID aren't 1000:1000 then volumes will break due to permission errors.

To handle this cleanly in most environments, you can make the UID and GID build time arguments but now we've entered a decently complex situation. I've written about and made videos about this topic at https://nickjanetakis.com/blog/running-docker-containers-as-a-non-root-user-with-a-custom-uid-and-gid. The blog post covers a Flask app but it applies to Rails in the same exact way. You can check out the Rails version of the Dockerfile at https://github.com/nickjj/docker-rails-example/blob/main/Dockerfile.

Basically it boils down to making UID / GID build args that default to `1000` and then you customize them if you happen to be on a system where you need to use a different value. For reference I've had at least 60 or 70 folks who took my Docker course mention they needed to customize this value because things didn't work for them out of the box before I modified things to allow customizing these values.

I guess the case I'm trying to make here is we can't depend on everyone having a UID / GID of `1000` in production and if these build args weren't added by default, the `Dockerfile` modifications to support this will stop anyone dead in their tracks unless they are well versed with Docker already. This is a pretty low level topic with a mixture of Linux file permission knowledge and Docker specifics.

**In CI**, yeah this is out of our control. I often disable all volumes in CI because if UID / GIDs are build time args, you're building images for production in CI in which case you want to use your prod environment's UID / GID values, not CI. It's just easier to disable all volumes in CI to make everything work. In practice this works out very well.

> Again. I'm open to -slim images, but it requires a bit of research first to see what common packages may be missing. e.g. trying to install some common open source apps to see what they depend on. Just pg and mysql likely isn't enough.

Some research was done and reported in this issue. It's listed in the "A couple of packages you may need to install" section. 

Beyond the gems listed there do you have additional gems you'd like to see tested? I don't mind doing the tests. You can find gems that require native dependencies to be compiled and try installing them in a `-slim` image. It will either work or throw an error at `bundle install` time related to missing a header file or some other dev dependency. The `build-essential` set of packages covers a lot of ground for common build tools.

> This Dockerfile is meant for deployment, not for development. So docker-compose doesn't make sense. The various datastores are expected to be provided by the hosting platform nad exposed via environment variables.

I think this comes down to what your goals are. As a contract worker I've seen a fair share of app deployments over the years. There's folks out there building million dollar businesses with single server deploys using Docker Compose where they run puma, sidekiq, action cable, postgres and redis in production through Docker Compose on a $40 / month server.

Docker Compose has gotten really good over the years if you're looking for a single server deployment solution. It's also really good for multi-server deploys as long as you're not talking about load balancing your app servers. Docker Compose v2 profiles make it painless to spin up your web, worker and cable processes on (3) different servers using the same `docker-compose.yml` file through setting a single `.env` value.

This approach can get you really far. Of course in that case once you go beyond (1) server you would be using a managed DB and Redis or at least have them being hosted separately from a specific server. In any case, using Docker Compose has real value in production.

**Comment by simi** [comment]
> The size of the base layers don't matter nearly as much as the size of the later layers assuming they don't change too often as you pointed. Trading 600MB in a base layer that only change monthly or so, for an extra 100MB into a later layer that is built much more often can lead to worse performance.

> Again, the overall size of an image isn't a good indication of performance. I can't help that there are morons on the internet...

Not all environments can benefit of layer caching. Good example are CI-likes envs starting from clean state everytime and pulling image in there would pull whole image size everytime. 1GB could be actually noticeable difference if you aim for fast CI build.

Anyway I do agree on keeping generated `Dockerfile` as simple as possible. But next to that one, maybe it would make sense to add "Docker" guide into Rails guides and also explain on some advanced best practices for advanced users like those mentioned by @nickjj.

**Comment by mfittko** [comment]
@nickjj we totally follow your objections here - specifically in regards of having totally bloated images for running standard rails applications for standard dependencies like `libvips`. We were bringing down image size to a mere fraction of the standard images and see huge benefits on our FARGATE based services. Btw, did you successfully end up with installing `build-essential` only during the bundle stage or any stage where it's required?

**Comment by nickjj** [comment]
@mfittko I usually keep them in. In my opinion it's not worth the battle to separate build time and runtime dependencies on a per environment basis. My example project is 500-600 MB in dev and prod which is ok enough. If you want to optimize it further you have to start making trade offs that require a decent amount of complexity, shaving off ~100-125 MB didn't seem worth it.

It is unfortunate EKS Fargate has such a high tax with container start up time since images aren't cached, that and each pod starts in its own world so you end up paying a 30-45s startup tax twice to run a migration and then your app in 2 different pods. I stopped using Fargate years ago due to that. I remember [commenting in this issue](https://github.com/aws/containers-roadmap/issues/696#issuecomment-984853159) about it.

**Comment by AlvesJorge** [comment]
Wow I'm really happy to see someone tackling such a similar issue as me.

I am wondering if you could explain to me how come installing a package (e.g: build-essential curl git) creates a 400mb difference? Obviously those packages shouldn't take that much space?

**Comment by nickjj** [comment]
build-essential has a bunch of packages rolled up together.

**Comment by AlvesJorge** [comment]
I was aware, but I still thought it had to be something else.
`apt remove build-essentials` frees up 20kb
running `apt autoremove` after frees up 278mb. 

I didn't think a package about _building code_ with _essentials_ would be this large.

Thanks!

**Comment by mfittko** [comment]
We're successfully using images without build-essential in the final build stage on production for a few months now and without an issue. That's most efficient if you have a working docker build cache setup, so also build stages are cached properly - otherwise you'll trade off some of the speed benefits for starting containers due to their smaller size for longer build times. But generally we're really happy to see like a 30% reduction in final image size! Rails web server image for a pretty heavy monolith is smaller than 500MB, which is okay if you add SOCI (on AWS).

**Comment by AlvesJorge** [comment]
Experimenting the same thing now!

**Comment by Ansarissab** [comment]
@nickjj Is it possible or a good/bad strategy if we use `node-slim` image and extract/copy the node's binary from there`/usr/local/bin/node` (just like we copy precompiled assets or node_modules) instead of installing it manually, just like you did in your `docker-rails-example` Dockerfile? Will it help reduce the image size? Thanks

**Comment by nickjj** [comment]
In the example Rails app I have there's a build stage that uses the Ruby image. It handles bundle installing gems + precompiling assets but the main app stage only copies in `/usr/local/bundle` and the precompiled finalized assets from that build stage. This way the main app stage avoids build time dependencies and the entire Node runtime. It gives you a fairly optimized image to run your app.
